### DISCUSSION TITLE
How to freeze Bullet simulation between two env.step ?

### DISCUSSION THREAD

**Discussion by charlypg** [upkie/upkie#425] [Software] [answered]
Hello, 
I would like to know if it is possible to freeze the Bullet simulation between to environment steps. 
In short, I want that 1 env.steps = nb-substeps in simulation, independantly from the time spent between two env.step but I did not find how to do so in the repo. 

The following example illustrates my point. In the time interval (1s) between two env.step the robot falls (simulation side). So that when the second env.step is called, the episode is terminated. But I don't want the robot's state to change between two env.step, as if only 1/frequency seconds had elapsed.

```python
import gymnasium as gym
import numpy as np
import upkie.envs
import time

upkie.envs.register()

with gym.make("UpkieGroundVelocity-v3", frequency=200.0) as env:
    observation, _ = env.reset()
    gain = np.array([10.0, 1.0, 0.0, 0.1])
    for step in range(1_000_000):
        action = gain.dot(observation).reshape((1,))
        observation, reward, terminated, truncated, _ = env.step(action)
        time.sleep(1.0) 
        if terminated or truncated:
            break 
```

Thank you in advance.

**Comment by stephane-caron** [comment, marked as answer]
For this purpose you can run the simulation with the `--nb-substeps` parameters. For instance, if the simulation spine runs at 1,000 Hz (default) and the agent runs at 200 Hz as in your example, you should set `--nb-substeps 5`:

### Using the simulation script

```console
./start_simulation.sh --show --nb-substeps 5
```

I've just pushed this update in 11f4f93b25d06d457546ade6c5868b2a6cb18237 so make sure you `git pull` before trying this :point_up_2: 

### Under the hood: using Bazel

```console
$ ./tools/bazelisk run //spines:bullet_spine -- --show --nb-substeps 5
```

You can also check out all available options by:

```console
$ ./tools/bazelisk run //spines:bullet_spine -- --help
INFO: ...

Optional arguments:

-h, --help
    Print this help and exit.
--log-dir <path>
    Path to a directory for output logs.
--nb-substeps <k>
    Number of simulation steps per action. Makes spine pausing.
--shm-name <name>
    Name for IPC shared memory file.
--show
    Show the Bullet GUI.
--space
    No ground, no gravity, fly like an eagle!
--extra-urdf-path
    Load extra URDFs into the environment.
--spine-frequency <frequency>
    Spine frequency in Hertz (default: 1000 Hz).
--robot-variant <variant>
    Robot variant (default: '').
-v, --version
    Print out the spine version number.
```

**Reply by charlypg** [reply]
Thank you a lot !  :smiley: 
It worked out for me.
